main.go

```diff
@@ -103,81 +103,153 @@ func main() {
-			var secrets corev1.SecretList
-			err := client.List(context.Background(), k8s.AllNamespaces, &secrets)
 			if err != nil {
 				log.Error().Err(err)
+				continue
 			}
-			log.Info().Msgf("Cluster has %v secrets", len(secrets.Items))
+		}
+
+		// sleep random time around 900 seconds
+		sleepTime := applyJitter(900)
+		log.Info().Msgf("Sleeping for %v seconds...", sleepTime)
+		time.Sleep(time.Duration(sleepTime) * time.Second)
+	}
+}
 
-			// loop all secrets
-			for _, secret := range secrets.Items {
-				waitGroup.Add(1)
-				status, err := processSecret(client, secret, "poller")
-				certificateTotals.With(prometheus.Labels{"namespace": *secret.Metadata.Namespace, "status": status, "initiator": "poller", "type": "secret"}).Inc()
-				waitGroup.Done()
+func watchNamespaces(waitGroup *sync.WaitGroup, client *k8s.Client) {
+	// loop indefinitely
+	for {
+		log.Info().Msg("Watching for new namespaces...")
+		var namespace corev1.Namespace
+		watcher, err := client.Watch(context.Background(), k8s.AllNamespaces, &namespace, k8s.Timeout(time.Duration(300)*time.Second))
+		defer watcher.Close()
 
+		if err != nil {
+			log.Error().Err(err)
+		} else {
+			// loop indefinitely, unless it errors
+			for {
+				namespace := new(corev1.Namespace)
+				eventType, err := watcher.Next(namespace)
 				if err != nil {
 					log.Error().Err(err)
-					continue
+					break
 				}
-			}
 
-			// sleep random time around 900 seconds
-			sleepTime := applyJitter(900)
-			log.Info().Msgf("Sleeping for %v seconds...", sleepTime)
-			time.Sleep(time.Duration(sleepTime) * time.Second)
+				if eventType == k8s.EventAdded {
+					log.Info().Msg("Listing secrets with 'copyToAllNamespaces' for all namespaces...")
+					var secrets corev1.SecretList
+					err := client.List(context.Background(), k8s.AllNamespaces, &secrets)
+					if err != nil {
+						log.Error().Err(err)
+					}
+					log.Info().Msgf("Cluster has %v secrets, checking for secrets with 'copyToAllNamespaces'...", len(secrets.Items))
+
+					// loop all secrets
+					for _, secret := range secrets.Items {
+						copyToAllNamespacesValue, ok := secret.Metadata.Annotations[annotationLetsEncryptCertificateCopyToAllNamespaces]
+						if ok {
+							shouldCopyToAllNamespaces, err := strconv.ParseBool(copyToAllNamespacesValue)
+							if err != nil {
+								log.Error().Err(err)
+								continue
+							}
+							if shouldCopyToAllNamespaces {
+								waitGroup.Add(1)
+								err = copySecretToNamespace(client, secret, namespace, fmt.Sprintf("ns-watcher:%v", eventType))
+								waitGroup.Done()
+
+								if err != nil {
+									log.Error().Err(err)
+									continue
+								}
+							}
+						}
+					}
+
+					if err != nil {
+						log.Error().Err(err)
+						continue
+					}
+				}
+			}
 		}
-	}(waitGroup)
 
-	foundation.HandleGracefulShutdown(gracefulShutdown, waitGroup)
+		// sleep random time around 900 seconds
+		sleepTime := applyJitter(900)
```

Comment:
This should be shorter, since during this time the namespaces aren't watched (see other watcher). There it's 30 seconds which is still quite long, but it was probably to avoid the Kubernetes API from getting hammered when it errors.